main/autogen/autoloader.cc

```diff
@@ -0,0 +1,331 @@
+}
+
+AutoloaderConfig AutoloaderConfig::enterConfig(core::GlobalState &gs, const realmain::options::AutoloaderConfig &cfg) {
+    AutoloaderConfig out;
+    out.rootDir = cfg.rootDir;
+    out.preamble = cfg.preamble;
+    for (auto &str : cfg.modules) {
+        out.topLevelNamespaceRefs.emplace(gs.enterNameConstant(str));
+    }
+    for (auto &str : cfg.requireExcludes) {
+        out.excludedRequireRefs.emplace(gs.enterNameUTF8(str));
+    }
+    for (auto &nameParts : cfg.sameFileModules) {
+        vector<core::NameRef> refs;
+        for (auto &name : nameParts) {
+            refs.emplace_back(gs.enterNameConstant(name));
+        }
+        out.nonCollapsableModuleNames.emplace(refs);
+    }
+    out.absoluteIgnorePatterns = cfg.absoluteIgnorePatterns;
+    out.relativeIgnorePatterns = cfg.relativeIgnorePatterns;
+    return out;
+}
+
+NamedDefinition NamedDefinition::fromDef(core::Context ctx, ParsedFile &parsedFile, DefinitionRef def) {
+    vector<core::NameRef> parentName;
+    if (def.data(parsedFile).parent_ref.exists()) {
+        auto parentRef = def.data(parsedFile).parent_ref.data(parsedFile);
+        if (!parentRef.resolved.empty()) {
+            parentName = parentRef.resolved;
+        } else {
+            parentName = parentRef.name;
+        }
+    }
+    return {def.data(parsedFile), parsedFile.showFullName(ctx, def), parentName, parsedFile.requires,
+            parsedFile.tree.file};
+}
+
+void DefTree::show(core::Context ctx, int level) {
+    auto fileRefToString = [&](const NamedDefinition &nd) -> string_view { return nd.fileRef.data(ctx).path(); };
+    fmt::print("{} [{}]\n", name().show(ctx), fmt::map_join(namedDefs, ", ", fileRefToString));
+    for (auto &[name, tree] : children) {
+        for (int i = 0; i < level; ++i) {
+            fmt::print("  ");
+        }
+        tree->show(ctx, level + 1);
+    }
+}
+
+string DefTree::fullName(core::Context ctx) const {
+    return fmt::format("{}",
+                       fmt::map_join(nameParts, "::", [&](core::NameRef nr) -> string_view { return nr.show(ctx); }));
+}
+
+string join(string path, string file) {
+    if (file.empty()) {
+        return path;
+    }
+    return fmt::format("{}/{}", path, file);
+}
+
+bool visitDefTree(const DefTree &tree, std::function<bool(const DefTree &)> visit) {
+    bool descend = visit(tree);
+    if (!descend) {
+        return false;
+    }
+    for (const auto &[_, child] : tree.children) {
+        descend = visitDefTree(*child, visit);
+        if (!descend) {
+            return false;
+        }
+    }
+    return true;
+}
+
+core::FileRef DefTree::file() const {
+    core::FileRef ref;
+    if (!namedDefs.empty()) {
+        // TODO what if there are more than one?
+        ref = namedDefs[0].fileRef;
+    } else if (!nonBehaviorDefs.empty()) {
+        // TODO this lacks sorting from `definition_sort_key`
+        ref = nonBehaviorDefs[0].fileRef;
+    }
+    return ref;
+}
+
+static const core::FileRef EMPTY_FILE;
```

Comment:
what is this for? Do you need it? I think `f != EMPTY_FILE` is equivalent to `f.exists()`